sdks/java/harness/src/main/java/org/apache/beam/fn/harness/status/BeamFnStatusClient.java

```diff
@@ -0,0 +1,231 @@
+      this.instruction = instruction;
+      this.trackedThreadName = trackedThreadName;
+      this.timeSinceTransition = timeSinceTransition;
+    }
+  }
+
+  @VisibleForTesting
+  String getActiveProcessBundleState() {
+    StringJoiner activeBundlesState = new StringJoiner("\n");
+    activeBundlesState.add("========== ACTIVE PROCESSING BUNDLES ==========");
+    if (processBundleCache.getActiveBundleProcessors().isEmpty()) {
+      activeBundlesState.add("No active processing bundles.");
+    } else {
+      List<BundleState> bundleStates = new ArrayList<>();
+      processBundleCache.getActiveBundleProcessors().keySet().stream()
+          .forEach(
+              instruction -> {
+                BundleProcessor bundleProcessor = processBundleCache.find(instruction);
+                if (bundleProcessor != null) {
+                  ExecutionStateTracker executionStateTracker = bundleProcessor.getStateTracker();
+                  Thread trackedTread = executionStateTracker.getTrackedThread();
+                  if (trackedTread != null) {
+                    bundleStates.add(
+                        new BundleState(
+                            instruction,
+                            trackedTread.getName(),
+                            executionStateTracker.getMillisSinceLastTransition()));
+                  }
+                }
+              });
+      bundleStates.stream()
+          // reverse sort active bundle by time since last transition.
+          .sorted(Comparator.comparing(BundleState::getTimeSinceTransition).reversed())
+          .limit(10) // only keep top 10
+          .forEachOrdered(
+              bundleState -> {
+                activeBundlesState.add(
+                    String.format("---- Instruction %s ----", bundleState.getInstruction()));
+                activeBundlesState.add(
+                    String.format("Tracked thread: %s", bundleState.getTrackedThreadName()));
+                activeBundlesState.add(
+                    String.format(
+                        "Time since transition: %.2f seconds%n",
+                        bundleState.getTimeSinceTransition() / 1000.0));
+              });
+    }
+    return activeBundlesState.toString();
+  }
+
+  private class InboundObserver implements StreamObserver<BeamFnApi.WorkerStatusRequest> {
+    @Override
+    public void onNext(WorkerStatusRequest workerStatusRequest) {
+      StringJoiner status = new StringJoiner("\n");
+      status.add(getMemoryUsage());
+      status.add("\n");
+      status.add(getActiveProcessBundleState());
+      status.add("\n");
+      status.add(getThreadDump());
+      outboundObserver.onNext(
+          WorkerStatusResponse.newBuilder()
+              .setId(workerStatusRequest.getId())
+              .setStatusInfo(status.toString())
+              .build());
+    }
+
+    @Override
+    public void onError(Throwable t) {
+      LOG.error("Error getting SDK harness status", t);
+      outboundObserver.onError(t);
```

Comment:
I think outboundObserver.onError(t) will not actually send back the exception as an error response but instead shut down the grpc connection. We can consider either just logging the error without calling outboundObserver.onError(t), or try to catch exception in onNext() and send it back with
```
WorkerStatusResponse.newBuilder()
              .setId(workerStatusRequest.getId())
              .setError(exceptionMesasge)
              .build()
```

Reply:
If the onError has been called for the client, isn't it already shutting down so it's okay to terminate the connection?

Reply:
I think it would make sense to just log and not call onError on the outboundObserver